hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/tenant/OMTenantDeleteRequest.java

```diff
@@ -18,33 +18,260 @@
+
+    IOException exception = null;
+    OmVolumeArgs omVolumeArgs;
+    String volumeOwner = null;
+    // deleteVolume is true if volumeName is not empty string
+    boolean deleteVolume = volumeName.length() > 0;
+    OzoneManagerStorageProtos.PersistedUserVolumeInfo newVolumeList = null;
+
+    try {
+      if (deleteVolume) {
+        // check Acl
+        if (ozoneManager.getAclsEnabled()) {
+          checkAcls(ozoneManager, OzoneObj.ResourceType.VOLUME,
+              OzoneObj.StoreType.OZONE, IAccessAuthorizer.ACLType.DELETE,
+              volumeName, null, null);
+        }
+
+        acquiredVolumeLock = omMetadataManager.getLock().acquireWriteLock(
+            VOLUME_LOCK, volumeName);
+
+        omVolumeArgs = getVolumeInfo(omMetadataManager, volumeName);
+        // Check volume ref count
+        long volRefCount = omVolumeArgs.getRefCount();
+        if (volRefCount > 1L) {
+          LOG.warn("Volume '{}' has a greater than 1 reference count of " +
+                  "'{}'", volumeName, volRefCount);
+          throw new OMException("Volume '" + volumeName + "' has a greater " +
+              "than 1 reference count of '" + volRefCount + "'. This may " +
+              "indicate the volume is in use by some other Ozone features. " +
+              "Please disable such other features before trying to delete " +
+              "the tenant again.", VOLUME_IN_USE);
+        }
+
+        volumeOwner = omVolumeArgs.getOwnerName();
+        acquiredUserLock = omMetadataManager.getLock().acquireWriteLock(
+            USER_LOCK, volumeOwner);
+
+        // Check volume emptiness, again
+        if (!omMetadataManager.isVolumeEmpty(volumeName)) {
+          LOG.debug("volume: '{}' is not empty", volumeName);
+          throw new OMException("Aborting tenant deletion. " +
+              "Volume becomes non-empty somewhere between" +
+              "preExecute and validateAndUpdateCache", VOLUME_NOT_EMPTY);
+        }
+
+        // Actual volume deletion, follows OMVolumeDeleteRequest
+        newVolumeList =
+            omMetadataManager.getUserTable().get(volumeOwner);
+        newVolumeList = delVolumeFromOwnerList(newVolumeList, volumeName,
+            volumeOwner, transactionLogIndex);
+        final String dbUserKey = omMetadataManager.getUserKey(volumeOwner);
+        omMetadataManager.getUserTable().addCacheEntry(
+            new CacheKey<>(dbUserKey),
+            new CacheValue<>(Optional.of(newVolumeList), transactionLogIndex));
+        final String dbVolumeKey = omMetadataManager.getVolumeKey(volumeName);
+        omMetadataManager.getVolumeTable().addCacheEntry(
+            new CacheKey<>(dbVolumeKey),
+            new CacheValue<>(Optional.absent(), transactionLogIndex));
+
+        // TODO: Set response dbVolumeKey?
+      }
+
+      // TODO: Should hold some tenant lock here. Just in case !deleteVolume
+//      acquiredTenantLock = omMetadataManager.getLock().acquireWriteLock(
+//          TENANT_LOCK, tenantId);
```

Comment:
We haven't decided on whether we want the `TENANT_LOCK` or not yet. Last time I talked to @prashantpogde and we didn't think it is needed but it might become necessary here. Or a coarse-gained global-ish lock on the would also work as tenant creation/deletion is not intended to be a high-frequency operation (as long as the lock doesn't block other high-frequency requests).

Reply:
Ok, lets discuss offline and then we can update here with the result.

Reply:
Hi @errose28 , actually I have implemented `TENANT_LOCK`.

But there's the caveat of lock ordering (See `OzoneManagerLock` class header). I assigned weight `6` to `TENANT_LOCK`. Therefore, I have to release `TENANT_LOCK` before grabbing `VOLUME_LOCK` (weight `0`). This means there could be a gap where `TENANT_LOCK` is released but `VOLUME_LOCK` is not yet acquired. For example, here: 

https://github.com/apache/ozone/blob/882809889c79bd45ff908d9ccc312859861798c9/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/s3/tenant/OMTenantDeleteRequest.java#L152-L170

Not sure if this could cause any real issues. I think not?

1. Or sometimes I could first grab `VOLUME_LOCK`, then grab `TENANT_LOCK`, but that's not always viable, as shown in the above example. In this case it has to get volume name from tenant table before it could grab the `VOLUME_LOCK`.

2. Another solution might be to move `TENANT_LOCK` weight ahead of `VOLUME_LOCK`. i.e. making `TENANT_LOCK` weight `0`. Not sure we want to do that. We could discuss.

Reply:
The discussion concluded that `VOLUME_LOCK` would be used to protect tenant tables, with the exception that reading from `TenantStateTable` (or cache) will not be protected by any locks because entries in `TenantStateTable` won't be updated once it is written to by `CreateTenant` (i.e. no `UpdateTenant`).